### PR TITLE
Add `ForkedPromise::hasBranches()`.

### PR DESCRIPTION
**Pull request by kentonv**
I ended up not needing this, but figured I'd merge the code anyway in case it's useful in the future.

### CODE REVIEW

**Comment by kentonv** [comment]
OMG GCC what is this garbage.

```
In file included from src/kj/async-test.c++:23:
src/kj/async-test.c++: In member function ‘virtual void kj::{anonymous}::TestCase473::run()’:
src/kj/debug.h:221:5: error: this ‘for’ clause does not guard... [-Werror=misleading-indentation]
     for (::kj::_::Debug::Fault f(__FILE__, __LINE__, ::kj::Exception::Type::FAILED, \
     ^~~
src/kj/debug.h:335:19: note: in expansion of macro ‘KJ_REQUIRE’
 #define KJ_ASSERT KJ_REQUIRE
                   ^~~~~~~~~~
src/kj/async-test.c++:481:3: note: in expansion of macro ‘KJ_ASSERT’
   KJ_ASSERT(!fork.hasBranches());
   ^~~~~~~~~
src/kj/async-test.c++:482:3: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘for’
   {
   ^
In file included from src/kj/async-test.c++:23:
src/kj/debug.h:221:5: error: this ‘for’ clause does not guard... [-Werror=misleading-indentation]
     for (::kj::_::Debug::Fault f(__FILE__, __LINE__, ::kj::Exception::Type::FAILED, \
     ^~~
src/kj/debug.h:335:19: note: in expansion of macro ‘KJ_REQUIRE’
 #define KJ_ASSERT KJ_REQUIRE
                   ^~~~~~~~~~
src/kj/async-test.c++:497:3: note: in expansion of macro ‘KJ_ASSERT’
   KJ_ASSERT(fork.hasBranches());
   ^~~~~~~~~
src/kj/async-test.c++:499:3: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘for’
   {
   ^
```

File: c++/src/kj/async-test.c++ [hunk 1]
```diff
@@ -478,14 +478,23 @@ TEST(Async, Fork) {
 
   auto fork = promise.fork();
 
+  KJ_ASSERT(!fork.hasBranches());
```
Comment (kentonv) [inline c++/src/kj/async-test.c++:481]:
```suggestion
  KJ_ASSERT(!fork.hasBranches());

```

**Review by harrishancock — APPROVED** [review]

**Comment by harrishancock** [comment]
Check out this sweet prior art: https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/kj/async-io.c%2B%2B#L948-L952

File: c++/src/kj/async-test.c++ (hunk 1 above)
Comment (kentonv) [inline c++/src/kj/async-test.c++:481]:
```suggestion
#if __GNUC__ && !__clang__ && __GNUC__ >= 7
// GCC 7 decides the open-brace below is "misleadingly indented" as if it were guarded by the `for`
// that appears in the implementation of KJ_REQUIRE(). Shut up shut up shut up.
#pragma GCC diagnostic ignored "-Wmisleading-indentation"
#endif
    KJ_ASSERT(!fork.hasBranches());
```

**Pull request merged**